### PR TITLE
Feature get certificate and achievements list

### PR DESCRIPTION
**Pull request by purisaurabh**
This PR contains two api , first one Get Achievements for getting list of all achievements for the particular profile and another one is Get certificate for getting list of all certificates for the particular profile.

### CODE REVIEW

File: internal/api/handler/achievement_handler.go [hunk 1]
```diff
@@ -40,3 +41,25 @@ func CreateAchievementHandler(ctx context.Context, profileSvc service.Service) f
 		})
 	}
 }
+
+func GetAchievementsHandler(ctx context.Context, achSvc service.Service) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		profileID, err := helpers.GetParams(r)
```
Comment (OmkeshJosh) [inline internal/api/handler/achievement_handler.go:48]: Can we define new function like `GetParamByID("profile_id")` and get profileID

File: internal/api/handler/achievement_handler.go [hunk 2]
```diff
@@ -40,3 +41,25 @@ func CreateAchievementHandler(ctx context.Context, profileSvc service.Service) f
...
+		if err != nil {
+			middleware.ErrorResponse(w, http.StatusBadGateway, err)
+			zap.S().Error(err)
+			return
+		}
+
+		values, err := achSvc.GetAchievements(ctx, profileID)
```
Comment (OmkeshJosh) [inline internal/api/handler/achievement_handler.go]: achivementsResp, err := achSvc.GetAchievements(ctx, profileID)

File: internal/api/tests/achievement_handler_test.go [hunk 3]
```diff
@@ -100,3 +103,102 @@ func TestCreateAchievementHandler(t *testing.T) {
...
+	getAchievementHandler := handler.GetAchievementsHandler(context.Background(), achSvc)
+
+	tests := []struct {
+		name               string
+		queryParams        string
+		MockSvc            func(mockSvc *mocks.Service)
+		expectedStatusCode int
+	}{
+		{
+			name:        "Success for fetching achievements",
```
Comment (OmkeshJosh) [inline internal/api/tests/achievement_handler_test.go]: Success for fetching achievements -> success_achievement

File: internal/api/tests/achievement_handler_test.go [hunk 4]
```diff
@@ -100,3 +103,102 @@ func TestCreateAchievementHandler(t *testing.T) {
...
+						ProfileID:   1,
+						Name:        "Client Appreciation",
+						Description: "Description of Appreciation",
+					},
+				}, nil).Once()
+			},
+			expectedStatusCode: http.StatusOK,
+		},
+		{
+			name:        "Success for fetching multiple achievements",
```
Comment (OmkeshJosh) [inline internal/api/tests/achievement_handler_test.go]: Success for fetching achievements -> success_achievements

File: internal/app/service/achievement_service.go [hunk 5]
```diff
@@ -43,3 +45,19 @@ func (achSvc *service) CreateAchievement(ctx context.Context, cDetail dto.Create
 
 	return cDetail.ProfileID, nil
 }
+
+func (achSvc *service) GetAchievements(ctx context.Context, profileID string) (value []dto.AchievementResponse, err error) {
+	ID, err := helpers.ConvertStringToInt(profileID)
```
Comment (OmkeshJosh) [inline internal/app/service/achievement_service.go]: handle it into handler level string to int convert

File: internal/app/service/certificate_service.go [hunk 6]
```diff
@@ -5,13 +5,15 @@ import (
...
 	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/errors"
+	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/helpers"
 	"github.com/joshsoftware/profile_builder_backend_go/internal/repository"
 	"go.uber.org/zap"
 )
 
 // CertificateService represents a set of methods for accessing the certificates.
 type CertificateService interface {
 	CreateCertificate(ctx context.Context, cDetail dto.CreateCertificateRequest) (profileID int, err error)
+	GetCertificates(ctx context.Context, profileID string) (value []dto.CertificateResponse, err error)
```
Comment (OmkeshJosh) [inline internal/app/service/certificate_service.go]: try to keep same type vars for `profileID`

File: internal/app/service/certificate_service.go [hunk 7]
```diff
@@ -48,3 +50,19 @@ func (profileSvc *service) CreateCertificate(ctx context.Context, cDetail dto.Cr
 
 	return cDetail.ProfileID, nil
 }
+
+func (certificateSvc *service) GetCertificates(ctx context.Context, profileID string) (value []dto.CertificateResponse, err error) {
+	id, err := helpers.ConvertStringToInt(profileID)
```
Comment (OmkeshJosh) [inline internal/app/service/certificate_service.go]: same for this

File: internal/app/service/tests/achievement_service_test.go [hunk 8]
```diff
@@ -95,3 +97,85 @@ func TestCreateAchievement(t *testing.T) {
...
+			ProfileID: mockProfileId,
+			MockSetup: func(mockAchievementStorer *mocks.AchievementStorer, profileID string) {
+				profileIDInt, _ := strconv.Atoi(profileID)
+				mockAchievementStorer.On("GetAchievements", mock.Anything, profileIDInt).Return(mockResponseAchievement, nil).Once()
+			},
+			isErrorExpected: false,
+			wantResponse:    mockResponseAchievement,
+		},
+		{
+			Name:      "Fail get achievement",
```
Comment (OmkeshJosh) [inline internal/app/service/tests/achievement_service_test.go]: use `fail_get_achievement` and use same pattern for other places

File: internal/app/service/tests/certificate_service_test.go [hunk 9]
```diff
@@ -108,3 +110,83 @@ func TestCreateCertificate(t *testing.T) {
...
+			ProfileID: mockProfileID,
+			MockSetup: func(certMock *mocks.CertificateStorer, profileID string) {
+				profileIDInt, _ := strconv.Atoi(profileID)
+				certMock.On("GetCertificatesList", mock.Anything, profileIDInt).Return([]dto.CertificateResponse{}, errors.New("error")).Once()
+			},
+			isErrorExpected: true,
+			wantResponse:    []dto.CertificateResponse{},
+		},
+		{
+			Name:      "Success but no certificates",
```
Comment (OmkeshJosh) [inline internal/app/service/tests/certificate_service_test.go]: sucess_with_empty_resultset

File: internal/repository/certificate_repository.go [hunk 10]
```diff
@@ -62,3 +64,36 @@ func (profileStore *CertificateStore) CreateCertificate(ctx context.Context, val
 
 	return nil
 }
+
+// GetCertificates fetches certificates details from the database.
+func (certificateStore *CertificateStore) GetCertificatesList(ctx context.Context, profileID int) (values []dto.CertificateResponse, err error) {
```
Comment (OmkeshJosh) [inline internal/repository/certificate_repository.go]: `GetCertificatesList` is GET or List? it seems both.

as per your response type, lets make it `ListCertificates`

File: internal/repository/achievement_repository.go [hunk 11]
```diff
@@ -61,3 +63,35 @@ func (profileStore *AchievementStore) CreateAchievement(ctx context.Context, val
 
 	return nil
 }
+
+func (achStore *AchievementStore) GetAchievements(ctx context.Context, profileID int) (values []dto.AchievementResponse, err error) {
```
Comment (OmkeshJosh) [inline internal/repository/achievement_repository.go]: GetAchievements -> ListAchievements

File: internal/app/service/achievement_service.go [hunk 12]
```diff
@@ -5,13 +5,15 @@ import (
...
 	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/errors"
+	"github.com/joshsoftware/profile_builder_backend_go/internal/pkg/helpers"
 	"github.com/joshsoftware/profile_builder_backend_go/internal/repository"
 	"go.uber.org/zap"
 )
 
 // AchievementService represents a set of methods for accessing the achievements
 type AchievementService interface {
 	CreateAchievement(ctx context.Context, cDetail dto.CreateAchievementRequest) (profileID int, err error)
+	GetAchievements(ctx context.Context, profileID string) (value []dto.AchievementResponse, err error)
```
Comment (OmkeshJosh) [inline internal/app/service/achievement_service.go]: ListAchievements

Comment (OmkeshJosh) [inline internal/app/service/achievement_service.go]: Add ListFilter in method signature 

type ListFilter  struct {
   ActiveMentIDs []int
  Name []string
}

File: internal/api/handler/achievement_handler.go (hunk 2 above)
Comment (purisaurabh) [reply internal/api/handler/achievement_handler.go]: Done

File: internal/api/tests/achievement_handler_test.go (hunk 3 above)
Comment (purisaurabh) [reply internal/api/tests/achievement_handler_test.go]: Done

File: internal/api/tests/achievement_handler_test.go (hunk 4 above)
Comment (purisaurabh) [reply internal/api/tests/achievement_handler_test.go]: Done

File: internal/app/service/tests/achievement_service_test.go (hunk 8 above)
Comment (purisaurabh) [reply internal/app/service/tests/achievement_service_test.go]: kk

File: internal/app/service/certificate_service.go (hunk 7 above)
Comment (purisaurabh) [reply internal/app/service/certificate_service.go]: kk

File: internal/app/service/tests/certificate_service_test.go (hunk 9 above)
Comment (purisaurabh) [reply internal/app/service/tests/certificate_service_test.go]: done

File: internal/app/service/achievement_service.go (hunk 5 above)
Comment (purisaurabh) [reply internal/app/service/achievement_service.go]: done

File: internal/app/service/certificate_service.go (hunk 6 above)
Comment (purisaurabh) [reply internal/app/service/certificate_service.go]: kk

File: internal/app/service/achievement_service.go (hunk 12 above)
Comment (purisaurabh) [reply internal/app/service/achievement_service.go]: done

File: internal/repository/achievement_repository.go (hunk 11 above)
Comment (purisaurabh) [reply internal/repository/achievement_repository.go]: Done

File: internal/repository/certificate_repository.go (hunk 10 above)
Comment (purisaurabh) [reply internal/repository/certificate_repository.go]: kk Done

File: internal/api/handler/achievement_handler.go (hunk 1 above)
Comment (purisaurabh) [reply internal/api/handler/achievement_handler.go:48]: GetParam is already defined for  getting the profileID so can i rename the function as  GetParamByID?

Comment (OmkeshJosh) [reply internal/api/handler/achievement_handler.go:48]: lets discuss offline

Comment (purisaurabh) [reply internal/api/handler/achievement_handler.go:48]: kk

File: internal/app/service/achievement_service.go (hunk 12 above)
Comment (purisaurabh) [reply internal/app/service/achievement_service.go]: kk

**Pull request merged**